[DOC] Distributed tracing documentation (#2022)
strimzi/strimzi-kafka-operator: documentation/book/con-overview-distributed-tracing.adoc

When instrumented, applications generate __traces__ for individual transactions. Traces are composed of __spans__, which define specific units of work.

To simplify the instrumentation of producers, consumers, and Kafka Streams applications, {ProductName} includes the https://github.com/opentracing-contrib/java-kafka-client/blob/master/README.md[OpenTracing Apache Kafka Client Instrumentation^] library. This library is an example of a __tracer__.

NOTE: The OpenTracing project is merging with the OpenCensus project. The new, combined project is named {OpenTelemetryHome}. OpenTelemetry will provide compatibility for applications that are instrumented using the OpenTracing APIs.

Jaeger, a tracing system, is an implementation of the OpenTracing APIs used for monitoring and troubleshooting microservices-based distributed systems. It consists of four main components and provides client libraries for instrumenting applications. You can use the Jaeger user interface to visualize, query, filter, and analyze trace data.

.An example of a query in the Jaeger user interface

image:image_con-overview-distributed-tracing.png[Simple Jaeger query]

> == Distributed tracing support in {ProductName} clients

Reviewer: Clients would suggets consumers and producers. Which is not the case here.

```suggestion
== Distributed tracing support in {ProductName}
```

Author: Updated in the source.

In {ProductName}, distributed tracing is supported in:

* Kafka Connect (including Kafka Connect with Source2Image support)
* Mirror Maker 

> You enable and configure distributed tracing for these types of clients by setting template configuration properties in the relevant custom resource (for example, `KafkaConnect`).

Reviewer: ```suggestion
You enable and configure distributed tracing for these types of compoenents by setting template configuration properties in the relevant custom resource (for example, `KafkaConnect`).
```

Author: Updated in the source.

> To enable distributed tracing in producers, consumers, and Kafka Streams applications, you can instrument application code using the included OpenTracing Apache Kafka Client Instrumentation library. When instrumented, these clients generate traces for transactions (for example, when producing messages or writing offsets to the log).

Reviewer: ```suggestion
To enable distributed tracing in Kafka Producer, Consumer, and Streams API applications, you can instrument application code using the OpenTracing Apache Kafka Client Instrumentation library.
When instrumented, these clients generate traces for messages (for example, when producing messages or writing offsets to the log).
```

Author: Good edits, thanks.

// Module included in the following assemblies:
//
// assembly-distributed-tracing.adoc

[id='con-overview-distributed-tracing-{context}']
= Overview of distributed tracing in {ProductName}

> Distributed tracing allows developers and system administrators to track the progress of transactions between applications and microservices in a distributed system. This information is useful for monitoring application performance at the micro-level and investigating issues with target systems and end-user applications.

Reviewer 2: microservice-level?

Author: Well, they may not use micro-services! I'll just remove that description.

In {ProductName} and data streaming platforms in general, distributed tracing facilitates the end-to-end tracking of messages: from source systems to the Kafka cluster and then to target systems and applications.

> As an aspect of system observability, distributed tracing complements the metrics that are available to view in xref:assembly-metrics-setup-{context}[Grafana dashboards], as well as the available loggers for each component. 

Reviewer 2: Observability. Can we use that term? I wonder if we need the subordinate clause here?

Author: I'm going to leave observability in there, as it seems to have some currency. Removed the subordinate clause.

.Introduction to OpenTracing and Jaeger

Distributed tracing in {ProductName} is implemented using the open source {OpenTracingHome} and {JaegerHome} projects.

> The OpenTracing specification defines APIs that developers can use to instrument applications -- often in microservice architectures -- for distributed tracing. It is independent from the tracing system.

Reviewer 2: often in microservice architectures. But not exclusively? Does that have any impact on where we say 'track the progress of transactions between applications and microservices in a distributed system'?

Author: Not exclusively, to my knowledge. Good point. Best to just remove this clause here and clarify what we mean in the first para:

"Distributed tracing allows developers and system administrators to track the progress of transactions between applications (and services in a microservice architecture) in a distributed system."

When instrumented, applications generate __traces__ for individual transactions. Traces are composed of __spans__, which define specific units of work.

> To simplify the instrumentation of Kafka Producer, Consumer, and Streams API applications, {ProductName} includes the https://github.com/opentracing-contrib/java-kafka-client/blob/master/README.md[OpenTracing Apache Kafka Client Instrumentation^] library. This library is an example of a __tracer__.

Reviewer 2: This library is an example of a __tracer__ implementation?

Author: I'm going to remove the sentence. I think the tracer is actually Jaeger.

NOTE: The OpenTracing project is merging with the OpenCensus project. The new, combined project is named {OpenTelemetryHome}. OpenTelemetry will provide compatibility for applications that are instrumented using the OpenTracing APIs.

Jaeger, a tracing system, is an implementation of the OpenTracing APIs used for monitoring and troubleshooting microservices-based distributed systems. It consists of four main components and provides client libraries for instrumenting applications. You can use the Jaeger user interface to visualize, query, filter, and analyze trace data.

.An example of a query in the Jaeger user interface

image:image_con-overview-distributed-tracing.png[Simple Jaeger query]

== Distributed tracing support in {ProductName}

In {ProductName}, distributed tracing is supported in:

* Kafka Connect (including Kafka Connect with Source2Image support)
* Mirror Maker 

> You enable and configure distributed tracing for these types of components by setting template configuration properties in the relevant custom resource (for example, `KafkaConnect`).

Reviewer 2: ```suggestion
You enable and configure distributed tracing for these components by setting template configuration properties in the relevant custom resource (for example, `KafkaConnect`).
```

To enable distributed tracing in Kafka Producer, Consumer, and Streams API applications, you can instrument application code using the OpenTracing Apache Kafka Client Instrumentation library.
When instrumented, these clients generate traces for messages (for example, when producing messages or writing offsets to the log).

Traces are sampled according to a sampling strategy and then visualized in the Jaeger user interface. This trace data is useful for monitoring the performance of your Kafka cluster and debugging issues with target systems and applications.

Before instrumenting a client application, you must first initialize and configure a Jaeger tracer in the Kafka cluster for use by the application, as described in xref:proc-configuring-jaeger-tracer-kafka-clients-{context}[Initializing a Jaeger tracer for Kafka clients].

> NOTE: Distributed tracing is not supported for Kafka brokers.

Reviewer 2: I wonder if this should be at the assembly level?

Author: Agreed, it works better there.
